doc/designs/quic-design/quic-io-arch.md

```diff
@@ -0,0 +1,485 @@
+libssl API allows an application to pass an arbitrary BIO to an SSL object; not
+only that, separate BIOs can be passed for the read and write directions. The
+nature of this BIO can be arbitrary; it could be a socket, or a memory buffer.
+
+Implementation of QUIC will require that the underlying network BIO passed to
+the QUIC implementation be configured to support datagram semantics instead of
+bytestream semantics as has been the case with traditional TLS over TCP.
+
+Implementation of QUIC requires handling of timer events as well as the
+circumstances where a network socket becomes readable or writable. In many cases
+we need to handle these events simultaneously (e.g. wait until a socket becomes
+readable, or a timeout expires, whichever comes first).
+
+Blocking vs. Non-Blocking I/O
+-----------------------------
+
+The above constraints make it effectively a requirement that non-blocking I/O be
+used for the calls to the underlying network BIOs. To illustrate this point, we
+first consider how QUIC might be implemented using blocking I/O internally.
+
+To function correctly and provide blocking semantics at the application level,
+our QUIC implementation must be able to block such that it can respond to any of
+the following events for the underlying network read and write BIOs immediately:
+
+- The underlying network write BIO becomes writeable;
+- The underlying network read BIO becomes readable;
+- A timeout expires.
+
+### Blocking sockets and select(3)
+
+Firstly, consider how this might be accomplished using the Berkeley sockets API.
+Blocking on all three wakeup conditions listed above would require use of an API
+such as select(3) or poll(3), regardless of whether the network socket is
+configured in blocking mode or not.
+
+While in principle APIs such as select(3) can be used with a socket in blocking
+mode, this is not an advisable usage mode. If a socket is in blocking mode,
+calls to send(3) or recv(3) may block for some arbitrary period of time, meaning
+that our QUIC implementation cannot handle incoming data (if we are blocked on
+send), send outgoing data (if we are blocked on receive), or handle timeout
+events.
+
+Though it can be argued that a select(3) call indicating readability or
+writeability should guarantee that a subsequent send(3) or recv(3) call will not
+block, there are several reasons why this is an extremely undesirable solution:
+
+- It is quite likely that there are buggy OSes out there which perform spurious
+  wakeups from select(3).
+
+- The fact that a socket is writeable does not necessarily mean that a datagram
+  of the size we wish to send is writeable, so a send(3) call could block
+  anyway.
+
+- This usage pattern precludes multithreaded use barring some locking scheme
+  due to the possibility of other threads racing between the call to select(3)
+  and the subsequent I/O call. This undermines our intentions to support
+  multi-threaded network I/O on the backend.
```

Comment:
But the API-facing architectural decisions we take now influence what we can do in the future. Rather confused about this since I was under the impression it was the entire reason we spent time developing `BIO_sendmmsg`.

Reply:
We can't have one thread per socket because we need to do reads and writes (and timeouts) simultaneously, hence @mattcaswell proposing we have multiple background threads, one making BIO_read calls to an application's (potentially custom) BIO and one making BIO_write calls. But our existing BIO API contract doesn't provide for concurrent use of BIO_read and BIO_write on the same BIO (not even our own BIO_s_datagram is safe under these conditions), so an application would have to make major changes to make its custom BIO concurrent to support this. The scale of those changes imposed upon an application seems to be to be a lot more major than just mandating non-blocking operation (as this document explains).

Reply:
But don't we require that any custom BIO implement support for BIO_sendmmsg() and BIO_recvmmsg()? Those APIs are defined to be thread safe....so this restriction applies even in the non-blocking scenario.